third_party/move/move-vm/types/src/values/values_impl.rs

```diff
@@ -1002,9 +1002,72 @@ impl IndexedRef {
     fn read_ref(self) -> PartialVMResult<Value> {
         use Container::*;
 
-        let res = match self.container_ref.container() {
-            Vec(r) => r.borrow()[self.idx].copy_value()?,
-            Struct(r) => r.borrow()[self.idx].copy_value()?,
+        Ok(Value(match self.container_ref.container() {
+            Vec(r) => {
+                let value = r.borrow()[self.idx].copy_value()?;
+
+                match value {
+                    // IndexedRef to vector must point to delayed fields or closures, because the
+                    // rest of primitive types are handled by specialized containers.
+                    ValueImpl::DelayedFieldID { .. } | ValueImpl::ClosureValue(_) => (),
+
+                    ValueImpl::Invalid
+                    | ValueImpl::U8(_)
+                    | ValueImpl::U16(_)
+                    | ValueImpl::U32(_)
+                    | ValueImpl::U64(_)
+                    | ValueImpl::U128(_)
+                    | ValueImpl::U256(_)
+                    | ValueImpl::Bool(_)
+                    | ValueImpl::Address(_)
+                    | ValueImpl::Container(_)
+                    | ValueImpl::ContainerRef(_)
+                    | ValueImpl::IndexedRef(_) => {
+                        return Err(PartialVMError::new(
+                            StatusCode::UNKNOWN_INVARIANT_VIOLATION_ERROR,
+                        )
+                        .with_message(format!(
+                            "invalid IndexedRef for vector container of {:?}",
+                            value
+                        )))
+                    },
+                }
+
+                value
+            },
+
+            Struct(r) | Locals(r) => {
+                let value = r.borrow()[self.idx].copy_value()?;
+
+                // IndexedRef of a struct or locals point to primitive types.
+                match value {
+                    ValueImpl::U8(_)
```

Comment:
I feel like we need basic APIs for determining predicates like primitive types, and some property of DelayedFieldId and ClosureValue that makes them the only one with IndexedRef, etc. Then we can just call those here.
If useful, APIs can be designed as filter (Some of itself or None), but not sure.

Reply:
Agree, lets do this, this match appears multiple times in this PR.

Reply:
Added a new `ValueKind` to get rid of most of matches